File: dotnet/FrameworkTests/ConnectorFacadeTests.cs
/*
 *  ====================
 *  DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
 *  
 *  Copyright 2008-2009 Sun Microsystems, Inc. All rights reserved.     
 *  
 *  The contents of this file are subject to the terms of the Common Development 
 *  and Distribution License("CDDL") (the "License").  You may not use this file 
 *  except in compliance with the License.
 *  
 *  You can obtain a copy of the License at 
 *  http://IdentityConnectors.dev.java.net/legal/license.txt
 *  See the License for the specific language governing permissions and limitations 
 *  under the License. 
 *  
 *  When distributing the Covered Code, include this CDDL Header Notice in each file
 *  and include the License file at identityconnectors/legal/license.txt.
 *  If applicable, add the following below this CDDL Header, with the fields 
 *  enclosed by brackets [] replaced by your own identifying information: 
 *  "Portions Copyrighted [year] [name of copyright owner]"
 *  ====================
 */

using System;
using System.Collections.Generic;

using NUnit.Framework;
using Org.IdentityConnectors.Common;
using Org.IdentityConnectors.Common.Security;
using Org.IdentityConnectors.Framework.Api;
using Org.IdentityConnectors.Framework.Api.Operations;
using Org.IdentityConnectors.Framework.Common;
using Org.IdentityConnectors.Framework.Common.Objects;
using Org.IdentityConnectors.Framework.Spi;
using Org.IdentityConnectors.Framework.Spi.Operations;
using Org.IdentityConnectors.Test.Common;

namespace FrameworkTests
{

    [TestFixture]
    public class ConnectorFacadeTests
    {

        [SetUp]
        public void setup()
        {
            // always reset the call patterns..
            MockConnector.Reset();
        }

        private class TestOperationPattern
        {
            /// <summary>
            /// Simple call back to make the 'facade' calls.
            /// </summary> 
            public Action<ConnectorFacade> MakeCall;

            /// <summary>
            /// Given the list of calls determine if they match expected values based
            /// on the calls made in the <see cref="TestOperationPattern.MakeCall"/> method.
            /// </summary>
            public Action<IList<MockConnector.Call>> CheckCalls;
        }

        /// <summary>
        /// Test the pattern of the common operations.
        /// </summary>
        private void TestCallPattern(TestOperationPattern pattern)
        {
            TestCallPattern(pattern, SafeType<Connector>.Get<MockAllOpsConnector>());
        }

        private void TestCallPattern(TestOperationPattern pattern,
                SafeType<Connector> clazz)
        {
            Configuration config = new MockConfiguration(false);
            ConnectorFacadeFactory factory = ConnectorFacadeFactory.GetInstance();
            // **test only**
            APIConfiguration impl = TestHelpers.CreateTestConfiguration(clazz, config);
            ConnectorFacade facade;
            facade = factory.NewInstance(impl);
            // make the call on the connector facade..
            pattern.MakeCall(facade);
            // check the call structure..
            IList<MockConnector.Call> calls = MockConnector.GetCallPattern();
            // check the call pattern..
            Assert.AreEqual("Init", calls[0].MethodName);
            calls.RemoveAt(0);
            pattern.CheckCalls(calls);
            Assert.AreEqual("Dispose", calls[0].MethodName);
            calls.RemoveAt(0);
            Assert.IsTrue(calls.Count == 0);
        }

        /// <summary>
        /// Tests that if an SPI operation is not implemented that the API will throw
        /// an <see cref="InvalidOperationException"/>.
        /// </summary>
        [Test]
        [ExpectedException(typeof(InvalidOperationException))]
        public void UnsupportedOperationTest()
        {
            Configuration config = new MockConfiguration(false);
            SafeType<Connector> clazz = SafeType<Connector>.Get<MockConnector>();
            ConnectorFacadeFactory factory = ConnectorFacadeFactory.GetInstance();
            APIConfiguration impl = TestHelpers.CreateTestConfiguration(clazz,
                    config);
            ConnectorFacade facade;
            facade = factory.NewInstance(impl);
            facade.Authenticate(ObjectClass.ACCOUNT, "fadf", new GuardedString(), null);
        }

        [Test]
        public void RunScriptOnConnectorCallPattern()
        {
            TestCallPattern(new TestOperationPattern()
            {
                MakeCall = facade =>
                {
                    facade.RunScriptOnConnector(
                            new ScriptContextBuilder("lang", "script").Build(),
                            null);
                },
                CheckCalls = calls =>
                {
                    Assert.AreEqual("RunScriptOnConnector", GetAndRemoveMethodName(calls));
                }
            });
        }

        [Test]
        public void RunScriptOnResourceCallPattern()
        {
            TestCallPattern(new TestOperationPattern()
            {
                MakeCall = facade =>
                {
                    facade.RunScriptOnResource(
                            new ScriptContextBuilder("lang", "script").Build(),
                            null);
                },
                CheckCalls = calls =>
                {
                    Assert.AreEqual("RunScriptOnResource", GetAndRemoveMethodName(calls));
                }
            });
        }

        /// <summary>
        /// Test the call pattern to get the schema.
        /// </summary>
        [Test]
        public void SchemaCallPattern()
        {
            TestCallPattern(new TestOperationPattern()
            {
                MakeCall = facade =>
                {
                    facade.Schema();
                },
                CheckCalls = calls =>
                {
                    Assert.AreEqual("Schema", GetAndRemoveMethodName(calls));
                }
            });
        }

        [Test]
        public void AuthenticateCallPattern()
        {
            TestCallPattern(new TestOperationPattern()
            {
                MakeCall = facade =>
                {
                    facade.Authenticate(ObjectClass.ACCOUNT, "dfadf", new GuardedString(), null);
                },
                CheckCalls = calls =>
                {
                    Assert.AreEqual("Authenticate", GetAndRemoveMethodName(calls));
                }
            });
        }

        [Test]
        public void ResolveUsernameCallPattern()
        {
            TestCallPattern(new TestOperationPattern()
            {
                MakeCall = facade =>
                {
                    facade.ResolveUsername(ObjectClass.ACCOUNT, "dfadf", null);
                },
                CheckCalls = calls =>
                {
                    Assert.AreEqual("ResolveUsername", GetAndRemoveMethodName(calls));
                }
            });
        }

        [Test]
        public void CreateCallPattern()
        {
            TestCallPattern(new TestOperationPattern()
            {
                MakeCall = facade =>
                {
                    ICollection<ConnectorAttribute> attrs = CollectionUtil.NewReadOnlySet<ConnectorAttribute>();
                    facade.Create(ObjectClass.ACCOUNT, attrs, null);
                },
                CheckCalls = calls =>
                {
                    Assert.AreEqual("Create", GetAndRemoveMethodName(calls));
                }
            });
        }

        [Test]
        [ExpectedException(typeof(ArgumentNullException))]
        public void CreateWithOutObjectClassPattern()
        {
            TestCallPattern(new TestOperationPattern()
            {
                MakeCall = facade =>
                {
                    ICollection<ConnectorAttribute> attrs = new HashSet<ConnectorAttribute>();
                    facade.Create(null, attrs, null);
                },
                CheckCalls = calls =>
                {
                    Assert.Fail("Should not get here..");
                }
            });
        }

        [Test]
        [ExpectedException(typeof(ArgumentException))]
        public void createDuplicatConnectorAttributesPattern()
        {
            TestCallPattern(new TestOperationPattern()
            {
                MakeCall = facade =>
                {
                    ICollection<ConnectorAttribute> attrs = new HashSet<ConnectorAttribute>();
                    attrs.Add(ConnectorAttributeBuilder.Build("abc", 1));
                    attrs.Add(ConnectorAttributeBuilder.Build("abc", 2));
                    facade.Create(ObjectClass.ACCOUNT, attrs, null);
                },
                CheckCalls = calls =>
                {
                    Assert.Fail("Should not get here..");
                }
            });
        }

        [Test]
        public void UpdateCallPattern()
        {
            TestCallPattern(new TestOperationPattern()
            {
                MakeCall = facade =>
                {
                    ICollection<ConnectorAttribute> attrs = new HashSet<ConnectorAttribute>();
                    attrs.Add(ConnectorAttributeBuilder.Build("accountid"));
                    facade.Update(ObjectClass.ACCOUNT, NewUid(0), attrs, null);
                },
                CheckCalls = calls =>
                {
                    Assert.AreEqual("Update", GetAndRemoveMethodName(calls));
                }
            });
        }

        [Test]
        public void DeleteCallPattern()
        {
            TestCallPattern(new TestOperationPattern()
            {
                MakeCall = facade =>
                {
                    facade.Delete(ObjectClass.ACCOUNT, NewUid(0), null);
                },
                CheckCalls = calls =>
                {
                    Assert.AreEqual("Delete", GetAndRemoveMethodName(calls));
                }
            });
        }

        [Test]
        public void SearchCallPattern()
        {
            TestCallPattern(new TestOperationPattern()
            {
                MakeCall = facade =>
                {
                    // create an empty results handler..
                    ResultsHandler rh = obj =>
                    {
                        return true;
                    };
                    // call the search method..
                    facade.Search(ObjectClass.ACCOUNT, null, rh, null);
                },
                CheckCalls = calls =>
                {
                    Assert.AreEqual("CreateFilterTranslator", GetAndRemoveMethodName(calls));
                    Assert.AreEqual("ExecuteQuery", GetAndRemoveMethodName(calls));
                }
            });
        }

        [Test]
        public void GetCallPattern()
        {
            TestCallPattern(new TestOperationPattern()
            {
                MakeCall = facade =>
                {
                    // create an empty results handler..
                    // call the search method..
                    facade.GetObject(ObjectClass.ACCOUNT, NewUid(0), null);
                },
                CheckCalls = calls =>
                {
                    Assert.AreEqual("CreateFilterTranslator", GetAndRemoveMethodName(calls));
                    Assert.AreEqual("ExecuteQuery", GetAndRemoveMethodName(calls));
                }
            });
        }

        [Test]
        public void TestOpCallPattern()
        {
            TestCallPattern(new TestOperationPattern()
            {
                MakeCall = facade =>
                {
                    facade.Test();
                },
                CheckCalls = calls =>
                {
                    Assert.AreEqual("Test", GetAndRemoveMethodName(calls));
                }
            });
        }

        [Test]
        public void UpdateMergeTests()
        {
            ConnectorAttribute expected, actual;
            Configuration config = new MockConfiguration(false);
            ConnectorFacadeFactory factory = ConnectorFacadeFactory.GetInstance();
            SafeType<Connector> clazz = SafeType<Connector>.Get<MockUpdateConnector>();
            // **test only**
            APIConfiguration impl = TestHelpers.CreateTestConfiguration(clazz, config);
            impl.SetTimeout(SafeType<APIOperation>.Get<GetApiOp>(), APIConstants.NO_TIMEOUT);
            impl.SetTimeout(SafeType<APIOperation>.Get<UpdateApiOp>(), APIConstants.NO_TIMEOUT);
            impl.SetTimeout(SafeType<APIOperation>.Get<SearchApiOp>(), APIConstants.NO_TIMEOUT);
            ConnectorFacade facade = factory.NewInstance(impl);
            // sniff test to make sure we can get an object..
            ConnectorObject obj = facade.GetObject(ObjectClass.ACCOUNT, NewUid(1), null);
            Assert.AreEqual(NewUid(1), obj.Uid);
            // ok lets add an attribute that doesn't exist..
            String ADDED = "somthing to add to the object";
            String ATTR_NAME = "added";
            ICollection<ConnectorAttribute> addAttrSet;
            addAttrSet = CollectionUtil.NewSet((IEnumerable<ConnectorAttribute>)obj.GetAttributes());
            addAttrSet.Add(ConnectorAttributeBuilder.Build(ATTR_NAME, ADDED));
            Name name = obj.Name;
            addAttrSet.Remove(name);
            Uid uid = facade.AddAttributeValues(ObjectClass.ACCOUNT, obj.Uid, ConnectorAttributeUtil.FilterUid(addAttrSet), null);
            // get back the object and see if there are the same..
            addAttrSet.Add(name);
            ConnectorObject addO = new ConnectorObject(ObjectClass.ACCOUNT, addAttrSet);
            obj = facade.GetObject(ObjectClass.ACCOUNT, NewUid(1), null);
            Assert.AreEqual(addO, obj);
            // attempt to add on to an existing attribute..
            addAttrSet.Remove(name);
            uid = facade.AddAttributeValues(ObjectClass.ACCOUNT, obj.Uid, ConnectorAttributeUtil.FilterUid(addAttrSet), null);
            // get the object back out and check on it..
            obj = facade.GetObject(ObjectClass.ACCOUNT, uid, null);
            expected = ConnectorAttributeBuilder.Build(ATTR_NAME, ADDED, ADDED);
            actual = obj.GetAttributeByName(ATTR_NAME);
            Assert.AreEqual(expected, actual);
            // attempt to delete a value from an attribute..
            ICollection<ConnectorAttribute> deleteAttrs = CollectionUtil.NewSet((IEnumerable<ConnectorAttribute>)addO.GetAttributes());
            deleteAttrs.Remove(name);
            uid = facade.RemoveAttributeValues(ObjectClass.ACCOUNT, addO.Uid, ConnectorAttributeUtil.FilterUid(deleteAttrs), null);
            obj = facade.GetObject(ObjectClass.ACCOUNT, uid, null);
            expected = ConnectorAttributeBuilder.Build(ATTR_NAME, ADDED);
            actual = obj.GetAttributeByName(ATTR_NAME);
            Assert.AreEqual(expected, actual);
            // attempt to delete an attribute that doesn't exist..
            ICollection<ConnectorAttribute> nonExist = new HashSet<ConnectorAttribute>();
            nonExist.Add(NewUid(1));
            nonExist.Add(ConnectorAttributeBuilder.Build("does not exist", "asdfe"));
            uid = facade.RemoveAttributeValues(ObjectClass.ACCOUNT, addO.Uid, ConnectorAttributeUtil.FilterUid(nonExist), null);
            obj = facade.GetObject(ObjectClass.ACCOUNT, NewUid(1), null);
            Assert.IsTrue(obj.GetAttributeByName("does not exist") == null);
        }

        static Uid NewUid(int id)
        {
            return new Uid(Convert.ToString(id));
        }

        static string GetAndRemoveMethodName(IList<MockConnector.Call> calls)
        {
            string result = calls[0].MethodName;
            calls.RemoveAt(0);
            return result;
        }
    }
}

File: connector-framework-internal/src/test/java/org/identityconnectors/framework/impl/api/ConnectorFacadeTests.java
/*
 * ====================
 * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS HEADER.
 *
 * Copyright 2008-2009 Sun Microsystems, Inc. All rights reserved.
 *
 * The contents of this file are subject to the terms of the Common Development
 * and Distribution License("CDDL") (the "License").  You may not use this file
 * except in compliance with the License.
 *
 * You can obtain a copy of the License at
 * http://opensource.org/licenses/cddl1.php
 * See the License for the specific language governing permissions and limitations
 * under the License.
 *
 * When distributing the Covered Code, include this CDDL Header Notice in each file
 * and include the License file at http://opensource.org/licenses/cddl1.php.
 * If applicable, add the following below this CDDL Header, with the fields
 * enclosed by brackets [] replaced by your own identifying information:
 * "Portions Copyrighted [year] [name of copyright owner]"
 * ====================
 */
package org.identityconnectors.framework.impl.api;

import static org.identityconnectors.framework.common.objects.ObjectClass.ACCOUNT;
import static org.testng.Assert.assertEquals;
import static org.testng.Assert.assertTrue;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import org.identityconnectors.common.CollectionUtil;
import org.identityconnectors.common.security.GuardedString;
import org.identityconnectors.framework.api.APIConfiguration;
import org.identityconnectors.framework.api.ConnectorFacade;
import org.identityconnectors.framework.api.ConnectorFacadeFactory;
import org.identityconnectors.framework.api.operations.APIOperation;
import org.identityconnectors.framework.api.operations.GetApiOp;
import org.identityconnectors.framework.api.operations.SearchApiOp;
import org.identityconnectors.framework.api.operations.UpdateApiOp;
import org.identityconnectors.framework.common.objects.Attribute;
import org.identityconnectors.framework.common.objects.AttributeBuilder;
import org.identityconnectors.framework.common.objects.AttributeUtil;
import org.identityconnectors.framework.common.objects.ConnectorObject;
import org.identityconnectors.framework.common.objects.Name;
import org.identityconnectors.framework.common.objects.ObjectClass;
import org.identityconnectors.framework.common.objects.ResultsHandler;
import org.identityconnectors.framework.common.objects.ScriptContextBuilder;
import org.identityconnectors.framework.common.objects.Uid;
import org.identityconnectors.framework.spi.Configuration;
import org.identityconnectors.framework.spi.Connector;
import org.identityconnectors.mockconnector.MockAllOpsConnector;
import org.identityconnectors.mockconnector.MockConfiguration;
import org.identityconnectors.mockconnector.MockConnector;
import org.identityconnectors.mockconnector.MockConnector.Call;
import org.identityconnectors.mockconnector.MockUpdateConnector;
import org.identityconnectors.test.common.TestHelpers;
import static org.testng.Assert.fail;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

public class ConnectorFacadeTests {

    // =======================================================================
    // Setup/Tear down
    // =======================================================================
    @BeforeMethod
    public void setup() {
        // always reset the call patterns..
        MockConnector.reset();
    }

    // =======================================================================
    // Helper Methods
    // =======================================================================
    private interface TestOperationPattern {
        /**
         * Simple call back to make the 'facade' calls.
         */
        public void makeCall(ConnectorFacade facade);

        /**
         * Given the list of calls determine if they match expected values based
         * on the calls made in the {@link #makeCall(ConnectorFacade)} method.
         */
        public void checkCalls(List<MockConnector.Call> calls);
    }

    /**
     * Test the pattern of the common operations.
     *
     * @throws ClassNotFoundException
     */
    @Test(enabled = false)
    private void testCallPattern(TestOperationPattern pattern) {
        testCallPattern(pattern, MockAllOpsConnector.class);
    }

    @Test(enabled = false)
    private void testCallPattern(TestOperationPattern pattern, Class<? extends Connector> clazz) {
        Configuration config = new MockConfiguration(false);
        ConnectorFacadeFactory factory = ConnectorFacadeFactory.getInstance();
        // **test only**
        APIConfiguration impl = TestHelpers.createTestConfiguration(clazz, config);
        ConnectorFacade facade;
        facade = factory.newInstance(impl);
        // make the call on the connector facade..
        pattern.makeCall(facade);
        // check the call structure..
        List<MockConnector.Call> calls = MockConnector.getCallPattern();
        // check the call pattern..
        assertEquals(calls.remove(0).getMethodName(), "init");
        pattern.checkCalls(calls);
        assertEquals(calls.remove(0).getMethodName(), "dispose");
        assertTrue(calls.isEmpty());
    }

    // =======================================================================
    // Tests
    // =======================================================================
    /**
     * Tests that if an SPI operation is not implemented that the API will throw
     * an {@link UnsupportedOperationException}.
     */
    @Test(expectedExceptions = UnsupportedOperationException.class)
    public void unsupportedOperationTest() {
        Configuration config = new MockConfiguration(false);
        Class<? extends Connector> clazz = MockConnector.class;
        ConnectorFacadeFactory factory = ConnectorFacadeFactory.getInstance();
        APIConfiguration impl = TestHelpers.createTestConfiguration(clazz, config);
        ConnectorFacade facade;
        facade = factory.newInstance(impl);
        facade.authenticate(ObjectClass.ACCOUNT, "fadf", new GuardedString("fadsf".toCharArray()),
                null);
    }

    @Test
    public void runScriptOnConnectorCallPattern() {
        testCallPattern(new TestOperationPattern() {
            public void makeCall(ConnectorFacade facade) {
                facade.runScriptOnConnector(new ScriptContextBuilder("lang", "script").build(),
                        null);
            }

            public void checkCalls(List<Call> calls) {
                assertEquals(calls.remove(0).getMethodName(), "runScriptOnConnector");
            }
        });
    }

    @Test
    public void runScriptOnResourceCallPattern() {
        testCallPattern(new TestOperationPattern() {
            public void makeCall(ConnectorFacade facade) {
                facade.runScriptOnResource(new ScriptContextBuilder("lang", "script").build(), null);
            }

            public void checkCalls(List<Call> calls) {
                assertEquals(calls.remove(0).getMethodName(), "runScriptOnResource");
            }
        });
    }

    /**
     * Test the call pattern to get the schema.
     */
    @Test
    public void schemaCallPattern() {
        testCallPattern(new TestOperationPattern() {
            public void makeCall(ConnectorFacade facade) {
                facade.schema();
            }

            public void checkCalls(List<Call> calls) {
                assertEquals(calls.remove(0).getMethodName(), "schema");
            }
        });
    }

    @Test
    public void authenticateCallPattern() {
        testCallPattern(new TestOperationPattern() {
            public void makeCall(ConnectorFacade facade) {
                facade.authenticate(ObjectClass.ACCOUNT, "dfadf", new GuardedString("fadfkj"
                        .toCharArray()), null);
            }

            public void checkCalls(List<Call> calls) {
                assertEquals(calls.remove(0).getMethodName(), "authenticate");
            }
        });
    }

    @Test
    public void resolveUsernameCallPattern() {
        testCallPattern(new TestOperationPattern() {
            public void makeCall(ConnectorFacade facade) {
                facade.resolveUsername(ObjectClass.ACCOUNT, "dfadf", null);
            }

            public void checkCalls(List<Call> calls) {
                assertEquals(calls.remove(0).getMethodName(), "resolveUsername");
            }
        });
    }

    @Test
    public void createCallPattern() {
        testCallPattern(new TestOperationPattern() {
            public void makeCall(ConnectorFacade facade) {
                Set<Attribute> attrs = CollectionUtil.<Attribute> newReadOnlySet();
                facade.create(ObjectClass.ACCOUNT, attrs, null);
            }

            public void checkCalls(List<Call> calls) {
                assertEquals(calls.remove(0).getMethodName(), "create");
            }
        });
    }

    @Test(expectedExceptions = NullPointerException.class)
    public void createWithOutObjectClassPattern() {
        testCallPattern(new TestOperationPattern() {
            public void makeCall(ConnectorFacade facade) {
                Set<Attribute> attrs = new HashSet<Attribute>();
                facade.create(null, attrs, null);
            }

            public void checkCalls(List<Call> calls) {
                fail("Should not get here..");
            }
        });
    }

    @Test(expectedExceptions = IllegalArgumentException.class)
    public void createDuplicatAttributesPattern() {
        testCallPattern(new TestOperationPattern() {
            public void makeCall(ConnectorFacade facade) {
                Set<Attribute> attrs = new HashSet<Attribute>();
                attrs.add(AttributeBuilder.build("abc", 1));
                attrs.add(AttributeBuilder.build("abc", 2));
                facade.create(ObjectClass.ACCOUNT, attrs, null);
            }

            public void checkCalls(List<Call> calls) {
                fail("Should not get here..");
            }
        });
    }

    @Test
    public void updateCallPattern() {
        testCallPattern(new TestOperationPattern() {
            public void makeCall(ConnectorFacade facade) {
                Set<Attribute> attrs = new HashSet<Attribute>();
                attrs.add(AttributeBuilder.build("accountid"));
                facade.update(ObjectClass.ACCOUNT, newUid(0), attrs, null);
            }

            public void checkCalls(List<Call> calls) {
                assertEquals(calls.remove(0).getMethodName(), "update");
            }
        });
    }

    @Test
    public void deleteCallPattern() {
        testCallPattern(new TestOperationPattern() {
            public void makeCall(ConnectorFacade facade) {
                facade.delete(ObjectClass.ACCOUNT, newUid(0), null);
            }

            public void checkCalls(List<Call> calls) {
                assertEquals(calls.remove(0).getMethodName(), "delete");
            }
        });
    }

    @Test
    public void searchCallPattern() {
        testCallPattern(new TestOperationPattern() {
            public void makeCall(ConnectorFacade facade) {
                // create an empty results handler..
                ResultsHandler rh = new ResultsHandler() {
                    public boolean handle(ConnectorObject obj) {
                        return true;
                    }
                };
                // call the search method..
                facade.search(ObjectClass.ACCOUNT, null, rh, null);
            }

            public void checkCalls(List<Call> calls) {
                assertEquals(calls.remove(0).getMethodName(), "createFilterTranslator");
                assertEquals(calls.remove(0).getMethodName(), "executeQuery");
            }
        });
    }

    @Test
    public void getCallPattern() {
        testCallPattern(new TestOperationPattern() {
            public void makeCall(ConnectorFacade facade) {
                // create an empty results handler..
                // call the search method..
                facade.getObject(ObjectClass.ACCOUNT, newUid(0), null);
            }

            public void checkCalls(List<Call> calls) {
                assertEquals(calls.remove(0).getMethodName(), "createFilterTranslator");
                assertEquals(calls.remove(0).getMethodName(), "executeQuery");
            }
        });
    }

    @Test
    public void testOpCallPattern() {
        testCallPattern(new TestOperationPattern() {
            public void makeCall(ConnectorFacade facade) {
                facade.test();
            }

            public void checkCalls(List<Call> calls) {
                assertEquals(calls.remove(0).getMethodName(), "test");
            }
        });
    }

    @Test
    public void updateMergeTests() {
        Attribute expected, actual;
        Configuration config = new MockConfiguration(false);
        ConnectorFacadeFactory factory = ConnectorFacadeFactory.getInstance();
        Class<? extends Connector> clazz = MockUpdateConnector.class;
        // **test only**
        APIConfiguration impl = TestHelpers.createTestConfiguration(clazz, config);
        impl.setTimeout(GetApiOp.class, APIOperation.NO_TIMEOUT);
        impl.setTimeout(UpdateApiOp.class, APIOperation.NO_TIMEOUT);
        impl.setTimeout(SearchApiOp.class, APIOperation.NO_TIMEOUT);
        ConnectorFacade facade = factory.newInstance(impl);
        // sniff test to make sure we can get an object..
        ConnectorObject obj = facade.getObject(ACCOUNT, newUid(1), null);
        assertEquals(obj.getUid(), newUid(1));
        // ok lets add an attribute that doesn't exist..
        final String ADDED = "somthing to add to the object";
        final String ATTR_NAME = "added";
        Set<Attribute> addAttrSet;
        addAttrSet = CollectionUtil.newSet(obj.getAttributes());
        addAttrSet.add(AttributeBuilder.build(ATTR_NAME, ADDED));
        Name name = obj.getName();
        addAttrSet.remove(name);
        Uid uid =
                facade.addAttributeValues(ACCOUNT, obj.getUid(), AttributeUtil
                        .filterUid(addAttrSet), null);
        // get back the object and see if there are the same..
        addAttrSet.add(name);
        ConnectorObject addO = new ConnectorObject(ACCOUNT, addAttrSet);
        obj = facade.getObject(ObjectClass.ACCOUNT, newUid(1), null);
        assertEquals(obj, addO);
        // attempt to add on to an existing attribute..
        addAttrSet.remove(name);
        uid =
                facade.addAttributeValues(ACCOUNT, obj.getUid(), AttributeUtil
                        .filterUid(addAttrSet), null);
        // get the object back out and check on it..
        obj = facade.getObject(ObjectClass.ACCOUNT, uid, null);
        expected = AttributeBuilder.build(ATTR_NAME, ADDED, ADDED);
        actual = obj.getAttributeByName(ATTR_NAME);
        assertEquals(actual, expected);
        // attempt to delete a value from an attribute..
        Set<Attribute> deleteAttrs = CollectionUtil.newSet(addO.getAttributes());
        deleteAttrs.remove(name);
        uid =
                facade.removeAttributeValues(ACCOUNT, addO.getUid(), AttributeUtil
                        .filterUid(deleteAttrs), null);
        obj = facade.getObject(ObjectClass.ACCOUNT, uid, null);
        expected = AttributeBuilder.build(ATTR_NAME, ADDED);
        actual = obj.getAttributeByName(ATTR_NAME);
        assertEquals(actual, expected);
        // attempt to delete an attribute that doesn't exist..
        Set<Attribute> nonExist = new HashSet<Attribute>();
        nonExist.add(newUid(1));
        nonExist.add(AttributeBuilder.build("does not exist", "asdfe"));
        uid =
                facade.removeAttributeValues(ACCOUNT, addO.getUid(), AttributeUtil
                        .filterUid(nonExist), null);
        obj = facade.getObject(ObjectClass.ACCOUNT, newUid(1), null);
        assertTrue(obj.getAttributeByName("does not exist") == null);
    }

    static Uid newUid(int id) {
        return new Uid(Integer.toString(id));
    }
}
